src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -48,60 +49,99 @@ internal static SafeChannelBindingHandle QueryChannelBinding(SafeSslHandle conte
 
         internal static SafeSslHandle AllocateSslContext(SslProtocols protocols, SafeX509Handle certHandle, SafeEvpPKeyHandle certKeyHandle, EncryptionPolicy policy, bool isServer, bool remoteCertRequired)
         {
-            SafeSslHandle context = null;
+            bool hasCertReference = false, 
+                hasCertKeyReference = false;
 
-            IntPtr method = GetSslMethod(protocols);
+            bool hasCertificateAndKey = 
+                certHandle != null && !certHandle.IsInvalid
+                && certKeyHandle != null && !certKeyHandle.IsInvalid;
 
-            using (SafeSslContextHandle innerContext = Ssl.SslCtxCreate(method))
+            try
             {
-                if (innerContext.IsInvalid)
+                if (hasCertificateAndKey)
                 {
-                    throw CreateSslException(SR.net_allocate_ssl_context_failed);
+                    certHandle.DangerousAddRef(ref hasCertReference);
+                    certKeyHandle.DangerousAddRef(ref hasCertKeyReference);
```

Comment:
I disagree.
a) We're talking about a failure case where safe handles have been incorrectly disposed of while still in use. Allocations that happen on the success path are fine to happen on this path.
b) This keeps the scoping tight so that the ref counts are only artificially raised for as brief a period of time as possible and we don't artificially keep handles alive that could otherwise be collected.
c) It helps keep the code cleaner, with less of it inside of an indented try block and less distance between where variables are declared and they're used.
d) There are some JIT benefits to code not being in a try block.